MetaData compatibility with frozen dataclasses

nstarman opened the pull request:
Using `setattr(instance, …)` calls the ``instance.__setattr__()`` method. However this can be ‘frozen’ like in `@dataclass(frozen=True)` and therefore fail. Changing to `object.__setattr__(instance, …)` is standard practice to ensure mutability in assignment.

mhvk reviewed (changes requested):
I don't think it makes sense for a descriptor to override the `__setattr__` of the instance it is on. If an instance is written to be immutable, then it should be actually immutable, which means that it either cannot use this descriptor, or it sets up it `__setattr__` such that `_meta` is special-cased. This is important to ensure proper hashing (usually one of the reasons to be immutable).

Now I can see that it would be handy to have `.meta` on an otherwise immutable class (even if that means one is no longer hashable), but the solution would seem to be to change the instance class. For instance, might it help to have a separate `AddMetaData` mixin class or so that brings in the descriptor and provides the relevant `__setattr__` for the instance?

nstarman commented:
https://docs.python.org/3/library/dataclasses.html#frozen-instances

Last sentence says to use this code pattern.

mhvk commented:
I still feel this is a problem to solve for the `dataclass`, not for the `descriptor` - something stuck on an instance should not be explicitly avoiding how an instance behaves (note that the last line in the link you give is consistent with that: it is about what the dataclass itself should do when it is frozen, not how a descriptor is meant to interact with it; from the python discussion I've seen, this is also something that really people would like to move away from, it is an obvious unpleasant hack...). 

Anyway, it still seems to be that a mixin class `DataClassMeta` is the answer here, not messing with `MetaData` itself -- it should not need to know or care what type of class it is attached to.

p.s. It is not so easy to judge the benefits of this PR without knowing why one would want to attach meta -- logically mutable -- to a frozen dataclass, which is not... I'm not convinced `Cosmology` is logical as a dataclass...

nstarman commented:
> I still feel this is a problem to solve for the `dataclass`, not for the `descriptor` - something stuck on an instance should not be explicitly avoiding how an instance behaves (note that the last line in the link you give is consistent with that: it is about what the dataclass itself should do when it is frozen, not how a descriptor is meant to interact with it; from the python discussion I've seen, this is also something that really people would like to move away from, it is an obvious unpleasant hack...).
> Anyway, it still seems to be that a mixin class `DataClassMeta` is the answer here, not messing with `MetaData` itself -- it should not need to know or care what type of class it is attached to.

Unpleasant and acknowledged to be "hacky", yes.  But also the best / only way for dataclasses since `@dataclass(frozen=True)` writes the `__setattr__` method for the class. 

![CleanShot 2023-10-01 at 11 54 01](https://github.com/astropy/astropy/assets/8949649/4d2a9c76-523d-4d5c-8672-7c7958a8258c)

This code pattern also closely matches the method by which `dataclass` itself sets attributes.
When `frozen=True`, `dataclass` uses `object.__setattr__` in the dynamically-generated `__init__`.
Descriptor-typed fields are supported (https://docs.python.org/3/library/dataclasses.html#descriptor-typed-fields).
When `frozen=True` this is literally the only way for descriptor fields to write to the object. `dataclass` then protects the the `Descriptor.__set__` thereafter, hiding it behind the frozen `object.__setattr__` and the fact that dataclass descriptors shouldn't return `self` and so are not directly accessible from the class.

So for any descriptor to interface with a frozen dataclass, something like this is the way to do it.


> p.s. It is not so easy to judge the benefits of this PR without knowing why one would want to attach meta -- logically mutable -- to a frozen dataclass, which is not...

This is definitely an issue, and I think a separate one from the above point. Consider this class:

![CleanShot 2023-10-01 at 12 02 32](https://github.com/astropy/astropy/assets/8949649/801fabd3-aa52-4deb-b73e-d6bf8e72c9e3)

This is a very pythonic construction, using `__slots__` to do all the freezing. Having mutable structures on "frozen" objects is AOK. I think this is a confusion of what is meant by "frozen". We have fully immutable structures where everything is frozen and we can have top-level frozen structures where `setattr` raises an error. So attaching `meta` to an object just means one is choosing the second form of frozen.

Here's a more verbose way of doing this with `__setattr__`, which is effectively how `@dataclass(frozen=True)` does it.

![CleanShot 2023-10-01 at 12 07 15](https://github.com/astropy/astropy/assets/8949649/a6f450b0-da7b-4539-97d1-f6ca80bf7c65)

nstarman commented:
I think our difference of opinion is because `dataclass` requires this construction, but it is as you point out, dangerous for non-dataclasses. `MetaData` can be used by both `dataclass` and non-`dataclass` so we should protect users and ensure that the different requirements are met for the difference uses.

mhvk commented:
Still don't like that a descriptor has to have any awareness of what it is bound too, but if it has to, it should indeed at least be explicit. Or we should allow the initialization to tell where to store the actual metadata; that at least properly separates the concerns (as would my suggested mixin class, or a `FrozenDataClassMetaData` class; explicit is better than implicit and all that).

nstarman commented:
Ok. I added the flag `obj_setattr` to `MetaData` so that it can support frozen as well as non-frozen dataclasses. `obj_setattr` only applies if the owner object is a dataclass. I've added tests and documentation.

nstarman commented:
Red to add a test for the obj setattr property.

nstarman commented:
> Still don't like that a descriptor has to have any awareness of what it is bound to

It's not necessary my favourite, but it has strong precedent: ``__set_name__`` is entirely to give a descriptor awareness of what it is bound to.

> it should indeed at least be explicit.

Agreed. That's why I added the explicit flag ``obj_setattr`` to turn on the stronger form of attribute setting.
Maybe we should call it `force_set`? I'm not strongly attached to the flag's name.

> Or we should allow the initialization to tell where to store the actual metadata; that at least properly separates the concerns

As in use ``__set_name__``  to choose the attribute name? That would be nice, but I think for a different PR as it doesn't address setting on a frozen dataclass.

> (as would my suggested mixin class, or a `FrozenDataClassMetaData` class; explicit is better than implicit and all that).

Unfortunately mixin classes are out, at least for dataclasses.
Metaclasses are maybe possible (IDK, [dataclasses are particular](https://peps.python.org/pep-0681/)) , but I don't think we need to go turn on that firehose for this spark of an issue.
IMO a simple flag, off by default, is good. Users making a `Table`, etc. will never see this flag! Not least of which is the it applies only to dataclasses but more so that it's set at the instantiation of MetaData, which occurs when the containing class is defined, not when the class is instantiated. So only the developer defining the class should ever encounter this.

nstarman commented:
Ping @mhvk. I think this solution is a simple fix to enable frozen dataclasses and only use ``object.__setattr__`` where strictly necessary.

nstarman commented on astropy/utils/metadata/core.py:
```diff
@@ -100,18 +149,25 @@ def __get__(self, instance, owner):
+    def _set_on_instance(self, instance, value):
+        """Set the meta attribute on the instance without copying."""
+        if is_dataclass(instance) and self._obj_setattr:
+            object.__setattr__(instance, "_meta", value)
```
Not sure why it says this isn't being tested! It most definitely is.

nstarman commented:
Ping @mhvk, I'm trying to get the Cosmology dataclass conversion in for v6.0 and this PR is one of a few in the pipeline. 
If this isn't a good solution, I can add a private `MetaData`-type object to `cosmology._utils`, but I would prefer to upgrade Astropy's general tool.

mhvk commented on astropy/utils/metadata/core.py:
```diff
@@ -100,18 +149,25 @@ def __get__(self, instance, owner):
         # to check for `None` everywhere.
         if value is None:
-            instance._meta = self.default_factory()
+            self._set_on_instance(instance, self.default_factory())
```
Looking at this, I feel that because of my mutterings, this has become too complicated...  Since `object.__setattr__` seems to be the standard for (frozen) dataclasses, how about just
```
if value is None:
    value = self.default_factory()
elif not ...
    ...

try:
    instance._meta = value
except AttributeError:
    if is_dataclass(instance) [and is_frozen??]:
        object.__setattr__(instance, "_meta", value)
    else:
        raise
```


And then get rid of the possible `obj_setattr`, etc.

nstarman replied:
Done! 
The ``is_dataclass(instance)`` check is so fast I avoided the `try-except`.

nstarman replied:
``is_dataclass`` is just ``hasattr(instance, '__dataclass_fields__')``.

mhvk reviewed (approved):
OK, definitely nice this way. Sorry for all the detours!

pllim commented:
We started seeing a new warning downstream at https://github.com/spacetelescope/acstools/issues/182 . Could it be caused by this and if so, it is supposed to happen? Hope you can advise. Thanks.

mhvk commented:
This should have zero impact unless frozen dataclasses are being used... @nstarman - do you have any idea?

nstarman commented:
There shouldn't be. There's no practical difference...
